main/server.go

```diff
@@ -0,0 +1,313 @@
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	var t todo.Todo
+	if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
+		return badRequest{err}
+	}
+	if t.ID == 0 {
+		t.ID = id
+	}
+	if t.ID != id {
+		return badRequest{fmt.Errorf("inconsistent todo IDs")}
+	}
+	if _, ok := todo.Get(c, id); !ok {
+		log.Infof(c, "Unable to find todo: %v", t)
+		return notFound{}
+	}
+	if err = t.Save(c); err != nil {
+		return err
+	}
+	return writeJson(w, http.StatusOK, t)
+}
+
+// DeleteTodo handles DELETE requests to /todos/{todoID}.
+// Returns a badRequest error if the ID cannot be parsed, and notFound if
+// no corresponding todo can be found.
+func DeleteTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	log.Infof(c, "Trying to delete id %v", id)
+	if ok := todo.Delete(c, id); !ok {
+		return notFound{}
+	}
+	w.WriteHeader(http.StatusNoContent)
+	return nil
+}
+
+// DeleteCompletedTodos handles DELETE requests to /todos.
+// It attempts to delete all todos which have been marked completed and returns
+// an error if one occurred.
+func DeleteCompletedTodos(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	if err := todo.DeleteCompleted(c); err != nil {
+		return err
+	}
+	w.WriteHeader(http.StatusNoContent)
+	return nil
+}
+
+// mergeJsonObjectFiles merges the k/v pairs in all indicated files with the
+// base map derived from the first file. An error is returned if any file
+// cannot be read or have its data unmarshaled. An error is also returned
+// if the resulting map cannot be marshaled.
+// Importantly, this function only works with files corresponding to JSON
+// objects! Files that correspond to JSON arrays will fail!
+func mergeJsonObjectFiles(files ...string) (b []byte, err error) {
+	if files == nil {
+		return nil, errors.New("Called with no files!")
+	}
+	b, err = ioutil.ReadFile(files[0])
+	if err != nil {
+		return nil, err
+	}
+	var base map[string]interface{}
+	if err = json.Unmarshal(b, &base); err != nil {
+		return nil, err
+	}
+	for _, file := range files[1:] {
+		b, err = ioutil.ReadFile(file)
+		if err != nil {
+			return nil, err
+		}
+		var next map[string]interface{}
+		if err = json.Unmarshal(b, &next); err != nil {
+			return nil, err
+		}
+		for k, v := range next {
+			base[k] = v
+		}
+	}
+	b, err = json.Marshal(base)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+// RenderSidebarJson writes a sidebar JSON file consumable by the TodoMVC
+// JS which is specific to our Go backend implementation.
+// On its first invocation, it will merge TodoMVC's relatively static
+// learn.json with our own backend data. The resultant []byte is then held
+// in memory for the lifetime of the server.
+func RenderSidebarJson(w http.ResponseWriter, r *http.Request) {
+	if learnJson == nil {
+		b, err := mergeJsonObjectFiles("static/todomvc/learn.json", "static/backend_learn.json")
```

Comment:
Is this an http template? I think just moving the body of this if into `init` should be enough.